bases/renku_data_services/background_jobs/utils.py

```diff
@@ -0,0 +1,54 @@
+"""Background jobs utilities."""
+
+import traceback
+from contextlib import AbstractAsyncContextManager
+from dataclasses import dataclass
+from types import TracebackType
+
+
+@dataclass
+class BackgroundJobError(Exception):
+    """Error raised when an exception happenend in a background job."""
+
+    message = "Background job failed:"
+    errors: list[BaseException]
+
+    def __format_errors__(self) -> str:
+        """Format contained errors for output."""
+        result = []
+        for error in self.errors:
+            result.append("".join(traceback.TracebackException.from_exception(error).format()))
+        return "\n".join(result)
```

Comment:
Out of curiosity: Do you have an example of how this appears in the logs? Hopefully reading multiple errors is feasible and not hard to process.

Reply:
certainly:
```
❯ python -m asyncio
asyncio REPL 3.12.5 (main, Aug 13 2024, 11:21:56) [GCC 14.2.1 20240805] on linux
Use "await" directly instead of "asyncio.run()".
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> from renku_data_services.background_jobs.utils import ErrorHandler
>>> eh = ErrorHandler()
>>> async with eh:
...     raise ValueError()
...
>>> async with eh:
...     raise OverflowError()
...
>>> eh.maybe_raise()
Traceback (most recent call last):
  File "/home/zenon/.pyenv/versions/3.12.5/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/zenon/.pyenv/versions/3.12.5/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/zenon/.pyenv/versions/3.12.5/lib/python3.12/asyncio/__main__.py", line 33, in callback
    coro = func()
           ^^^^^^
  File "<console>", line 1, in <module>
  File "/home/zenon/DEV/ETH/renku-data-services/bases/renku_data_services/background_jobs/utils.py", line 54, in maybe_raise
    raise BackgroundJobError(errors=self.errors)
renku_data_services.background_jobs.utils.BackgroundJobError: BackgroundJobError: Background job failed:
Traceback (most recent call last):
  File "<console>", line 2, in <module>
ValueError

Traceback (most recent call last):
  File "<console>", line 2, in <module>
OverflowError
```

Reply:
Thanks. It think it would be nice if we could have:
```
❯ python -m asyncio
asyncio REPL 3.12.5 (main, Aug 13 2024, 11:21:56) [GCC 14.2.1 20240805] on linux
Use "await" directly instead of "asyncio.run()".
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> from renku_data_services.background_jobs.utils import ErrorHandler
>>> eh = ErrorHandler()
>>> async with eh:
...     raise ValueError()
...
>>> async with eh:
...     raise OverflowError()
...
>>> eh.maybe_raise()
Traceback (most recent call last):
  File "/home/zenon/.pyenv/versions/3.12.5/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/home/zenon/.pyenv/versions/3.12.5/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/home/zenon/.pyenv/versions/3.12.5/lib/python3.12/asyncio/__main__.py", line 33, in callback
    coro = func()
           ^^^^^^
  File "<console>", line 1, in <module>
  File "/home/zenon/DEV/ETH/renku-data-services/bases/renku_data_services/background_jobs/utils.py", line 54, in maybe_raise
    raise BackgroundJobError(errors=self.errors)
renku_data_services.background_jobs.utils.BackgroundJobError: BackgroundJobError: Background job failed:
=== Error 1/2 ===
Traceback (most recent call last):
  File "<console>", line 2, in <module>
ValueError
=== Error 2/2 ===
Traceback (most recent call last):
  File "<console>", line 2, in <module>
OverflowError
```

Or similar. Do not see it as a blocking comment though.